### PR TITLE
Improve justification of Chinese/Japanese text

### PR DESCRIPTION
**Pull request by asayers**
Upstream:

image

The problem here is that typst is only allowed to insert justification spacing immediately after commas and periods.  However, the convention in Japanese typesetting is to insert spacing evenly between all characters.

This branch:

image

Much nicer!

Related to https://github.com/typst/typst/issues/276

### CODE REVIEW

File: library/src/text/shaping.rs [hunk 1]
```diff
@@ -69,8 +70,24 @@ impl ShapedGlyph {
...
+    }
+}
+
+/// Does this script separate its words using spaces?
+fn is_spaceless(script: Script) -> bool {
+    use Script::*;
+    match script {
+        Hiragana | Katakana | Han => true,
+        _ => false,
     }
```
Comment (laurmaedje) [inline library/src/text/shaping.rs:91]:
```suggestion
    use Script::*;
    matches!(script, Hiragana | Katakana | Han)
```

File: library/src/text/shaping.rs [hunk 2]
```diff
@@ -69,8 +70,24 @@ impl ShapedGlyph {
...
 
     /// Whether the glyph is justifiable.
+    ///
+    /// Typst's basic justification strategy is to stretch all the spaces
+    /// in a line until the line fills the available width.  However, some
+    /// scripts (notably Chinese and Japanese) don't use spaces.
+    ///
+    /// In Japanese typography, the convention is to insert space evenly
+    /// between all glyphs.  I assume it's the same in Chinese.
```
Comment (laurmaedje) [inline library/src/text/shaping.rs:79]:
```suggestion
    /// Typst's basic justification strategy is to stretch all the spaces
    /// in a line until the line fills the available width. However, some
    /// scripts (notably Chinese and Japanese) don't use spaces.
    ///
    /// In Japanese typography, the convention is to insert space evenly
    /// between all glyphs. I assume it's the same in Chinese.
```

**Comment by laurmaedje** [comment]
Makes sense!

**Comment by asayers** [comment]
Thanks!  And congratulations on doing such a good job with typst.  I thought we'd never see LaTeX dethroned, but now I think we might.

**Pull request merged**